File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    longest = 0  # the longest diameter 
    
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        
        # Returns height
        def dfs(curr):  
            if not curr:
                return 0
            
            left = dfs(curr.left)
            right = dfs(curr.right)
            self.longest = max(self.longest, left + right)  # Keep updating the diameter            
            return 1 + max(left, right)
        
        dfs(root)
        return self.longest
            

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {   
public:
    int diameterOfBinaryTree(TreeNode* root) {
        int longest = 0;
        dfs(root, longest);
        return longest;
    }
private:
    int dfs(TreeNode* node, int& longest) {
        if (!node) return 0;
        int left = dfs(node->left, longest);
        int right = dfs(node->right, longest);
        longest = max(longest, left + right);
        return 1 + max(left, right);
    }
};